### PR TITLE
Jupyter upgrades for exchange and support functions

### PR DESCRIPTION
**Pull request by cmickeyb**
Major update of jupyter notebooks:

 *  Update for the new jupyter functions in the python module.  Issuer and token issuer should be fully functional. Other notebooks are work in progress.
    
* New notebooks for wallets and orders. Do not expect these to work at present. More than placeholders but less than fully functional.
    
* Split the exchange launch page into some additional documentation files and the operational links to the factories.    
  
* Add Jupyter files for the root notebook directory. Replace the empty index file with one that contains somewhat useful content. Move a lot of the getting started content into its own file. This should be replaced later by notebooks that allow for managing configurations.

Major update for the python functions that support Jupyter notebooks:

* Create a new jupyter module that can be used for multiple contract families; this will include context handling and ipython magic functions.
    
* Update the context creation for exchange family templates; this should make context creation a little more configurable
    
* Split out the the reusable components of the exchange jupyter python modue into a separate module. This is both functions that should be exposed to a jupyter notebook and support functions that are note exposed.

### CODE REVIEW

File: exchange-contract/docs/notebooks/factories/wallet.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+#     display_name: Python 3 (ipykernel)
+#     language: python
+#     name: python3
+# ---
+
+# + [markdown] editable=true slideshow={"slide_type": ""}
+# # Wallet Factory
+#
+# Use this notebook to create a wallet associated with a particular issuer.
+
```
Comment (prakashngit) [inline exchange-contract/docs/notebooks/factories/wallet.py:19]: I am  afraid I do not understand which contract this notebook is going to interact with ? Is it a placeholder?

File: docs/jupyter.md [hunk 2]
```diff
@@ -17,6 +22,8 @@ virutal environment and that the Exchange contract family has been installed.
 * Activate the virtual environment: `source $PDO_INSTALL_ROOT/bin/activate`
 * Install the Jupyter notebook package: `pip install notebook`
 * Install [Papermill](https://papermill.readthedocs.io/en/latest/): `pip install papermill`
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply docs/jupyter.md:24]: in the docker folder.

File: docs/jupyter.md [hunk 3]
```diff
@@ -17,6 +22,8 @@ virutal environment and that the Exchange contract family has been installed.
...
+* Install [Jupytext](https://jupytext.readthedocs.io/en/latest/): `pip install jupytext`
+* Install [IPYWidgets](https://ipywidgets.readthedocs.io/en/stable/): `pip install ipywidgets`
 
 Note that you will need to reinstall Jupyter and papermill any time
 you rebuild the Python virtual environment.
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply docs/jupyter.md:29]: not sure i understand the question. this works. whether you want to run with docker (which complicates key management, notebook persistence, ... ) or bare metal (which requires the additional installation). so we don't "support" anything. we offer alternatives.

File: docs/notebooks/documents/getting_started.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    language: python
+    name: python3
+---
+
+# Getting Started with PDO Contracts #
+
+
+Assumptions: 
+* PDO client virtual environment has been installed in `${PDO_INSTALL_ROOT}`
+* PDO Contracts packages have been built and installed in the client virtual environment
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply docs/notebooks/documents/getting_started.md:20]: suggest changes. though i think you'll need a more significant re-write than just removing lines. this is clearly not in final form. for now, i consider a lot of these as placeholders to be improved with time. 
i do think a "WIP" to the title would be appropriate.

File: docs/notebooks/documents/getting_started.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+3. Create a service group configuration
+
+
+## Simplified configuration
+
+If you are running the PDO services in the common test configuration (e.g. with 5 provisioning services, 5 enclave service, and 5 storage services all running on the same host, you can create all of the necessary configuration files with the command:
+    
+```
+    $ export service_host=<service hostname>
+    $ ${PDO_INSTALL_ROOT}/opt/pdo/bin/pdo-create-service-groups.psh --service_host ${service_host}
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply docs/notebooks/documents/getting_started.md:34]: we're not pre-supposing container execution. and i don't think we should. yes, the container takes care of a few of these (though right now not well for an application user... they are really designed for us as developers).

File: docs/notebooks/documents/key_manager.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+#     display_name: Python 3 (ipykernel)
+#     language: python
+#     name: python3
+# ---
+
+# %% [markdown]
+# # Key Management for PDO Contracts
+#
+# This notebook helps to manage keys for signing PDO contract transactions.
+
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply docs/notebooks/documents/key_manager.py:19]: agreed. feature coming (with a few more like it)

File: exchange-contract/docs/notebooks/index.md [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+More detailed information about the exchange contract family is available in the
+[Exchange Contract Family Overview](documents/overview.ipynb).
+
+## Factories
+
+The links below connect you to "factories" for creating contract objects in the Exchange contract family.
+
+* [Create a new asset](factories/issuer.ipynb)
+* [Mint tokens for an asset](factories/token-issuer.ipynb)
+* [Create a wallet for assets](factories/wallet.ipynb)
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply exchange-contract/docs/notebooks/index.md:28]: suggest wordings.

File: exchange-contract/docs/notebooks/factories/wallet.py (hunk 1 above)
Comment (cmickeyb) [reply exchange-contract/docs/notebooks/factories/wallet.py:19]: placeholder. again, i need to label a bunch of these as WIP.

**Review by g2flyer — COMMENTED** [review]
feedback collected so far ...

File: docs/jupyter.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (cmickeyb) [reply docs/jupyter.md:29]: that's much clear. thank you. will update.

File: docs/notebooks/documents/getting_started.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+  kernelspec:
+    display_name: Python 3 (ipykernel)
+    language: python
+    name: python3
+---
+
+# Getting Started with PDO Contracts #
+
+
+Assumptions:
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply docs/notebooks/documents/getting_started.md:18]: i think this should be re-titled as instructions for getting started with host install. this content was copied (almost verbatim) from the old readme. ironically, no one cared about it there. :)

File: docs/notebooks/index.md [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+[Exchange Contract Family](exchange/index.ipynb)
+
+
+
+### Digital Assets
+
+
+The Digital Assets contract family implements a basic digital asset for bitmap images. The contracts extend the asset and token contracts in the Exchange contract family for sharing images with well-defined policies.
+
+
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply docs/notebooks/index.md:49]: done

File: exchange-contract/docs/notebooks/factories/wallet.py (hunk 1 above)
Comment (cmickeyb) [reply exchange-contract/docs/notebooks/factories/wallet.py:20]: Marked as WIP

File: exchange-contract/docs/notebooks/templates/token.py [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+# In the code block below, you will likely want to change the value of the export 
+# path to the directory where the contract family export file will be saved. Feel
+# free to change the file name as well. The default uses the asset name.
+
+# +
+# %%skip True
+export_file = '${{data}}/{}.zip'.format(token_class)
+
+contexts = ['asset_type', 'vetting', 'guardian', 'token_issuer', 'token_object']
+pc_jupyter.export_context_file(state, bindings, context, contexts, export_file)
```
[the remark this answers is not in the document]
Comment (cmickeyb) [reply exchange-contract/docs/notebooks/templates/token.py]: should be done, please check.

**Review by prakashngit — APPROVED** [review]

**Pull request merged**